src/xmerl_dsig.erl

```diff
@@ -0,0 +1,730 @@
+%%% OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+%%%
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+%%
+%% esaml - SAML for erlang
+%%
+%% Copyright (c) 2013, Alex Wilson and the University of Queensland
+%% All rights reserved.
+%%
+%% Distributed subject to the terms of the 2-clause BSD license, see
+%% the BSD2 file in the root of the distribution.
+
+%% @doc XML digital signatures for xmerl
+%%
+%% Functions for performing XML digital signature generation and
+%% verification, as specified at http://www.w3.org/TR/xmldsig-core/ .
+%%
+%% These routines work on xmerl data structures (see the xmerl user guide
+%% for details).
+%%
+%% Currently only RSA + SHA1|SHA256 signatures are supported, in the typical
+%% enveloped mode.
+-module(xmerl_dsig).
+
+-export([verify/1, verify/2, sign/3, strip/1, digest/1]).
+
+-include_lib("xmerl/include/xmerl.hrl").
+-include_lib("public_key/include/public_key.hrl").
+
+-type xml_thing() :: #xmlDocument{} | #xmlElement{} | #xmlAttribute{} | #xmlPI{} | #xmlText{} | #xmlComment{}.
+-type fingerprint() :: binary() | {sha | sha256, binary()}.
+
+%% @doc Returns an xmlelement without any ds:Signature elements that are inside it.
+-spec strip(Element :: #xmlElement{} | #xmlDocument{}) -> #xmlElement{}.
+strip(#xmlDocument{content = Kids} = Doc) ->
+    NewKids = [if (element(1,K) =:= xmlElement) -> strip(K); true -> K end || K <- Kids],
+    Doc#xmlDocument{content = NewKids};
+strip(#xmlElement{content = Kids} = Elem) ->
+    NewKids = [Kid || Kid <- Kids, is_valid_kid(Kid)],
+    Elem#xmlElement{content = NewKids}.
+
+is_valid_kid(Kid) when is_record(Kid, xmlAttribute); is_record(Kid, xmlElement) ->
+    Canon_Name = xmerl_c14n:canon_name(Kid),
+    Canon_Name =/= "http://www.w3.org/2000/09/xmldsig#Signature";
+is_valid_kid(_Child) -> true.
+
+strip_signature(Doc, #xmlElement{name = Name, parents = Parents}) ->
+    do_strip_signature(Doc, Name, lists:reverse(Parents)).
+
+do_strip_signature(#xmlElement{name = Match, content = Kids} = Elem, Name,
+                   [{Match, _Pos} | Parents]) ->
+    logger:debug("Match: ~p~n", [Match]),
+    NewKids = do_strip_signature(Kids, Name, Parents),
+    Elem#xmlElement{content = NewKids};
+do_strip_signature([#xmlElement{name = Signature} | Tail], Signature, _) ->
+    logger:debug("Signature: ~p~n~n", [Signature]),
+    Tail;
+do_strip_signature([#xmlElement{name = Match, content = Kids} = Elem | Tail], Name,
+                   [{Match, _Pos} | Parents]) ->
+    logger:debug("Match: ~p~n", [Match]),
+    NewKids = do_strip_signature(Kids, Name, Parents),
+    [Elem#xmlElement{content = NewKids} | Tail];
+do_strip_signature([#xmlElement{name = _NoMatch} = Elem | Tail], Name,
+                   Parents) ->
+    [Elem | do_strip_signature(Tail, Name, Parents)];
+do_strip_signature([NotElem | Tail], Name, Parents) ->
+    [NotElem | do_strip_signature(Tail, Name, Parents)];
+do_strip_signature([], _, _) ->
+    [].
```

Comment:
`strip` and `strip_signature` is an unfortunate name clash. They sound like exactly the same thing, especially given the @doc comment for `strip`. There is also `strip_it`.

If some of these are only for working around application specific issues they should be extracted from this library. If they are not they need better names and a comment each explaining what their purpose is.